tests/main/userns/task.yaml

```
@@ -0,0 +1,120 @@
+    # be the case when using the vendored apparmor
+    if snap debug sandbox-features --required apparmor:parser:snapd-internal; then
+      if ! snap debug sandbox-features --required apparmor:parser:userns; then
+        echo "The internal vendored apparmor within snapd does not support userns restrictions"
+        exit 1
+      fi
+    else
+      if ! snap debug sandbox-features --required apparmor:parser:userns; then
+        # apparmor in ubuntu since kinetic should support userns
+        if os.query is-ubuntu-ge 22.10; then
+          echo "Ubuntu host system AppArmor does not support userns but it should"
+          exit 1
+        fi
+        echo "Host system AppArmor does not support userns, skipping remainder of test"
+        exit 0
+      fi
+    fi
+
+    # now check the kernel AppArmor userns enforcement as well as seccomp
+    # enforcement above
+    if [ ! -e /proc/sys/kernel/apparmor_restrict_unprivileged_userns ]; then
+      echo "Skipping testing of AppArmor enforcement since it is not supported by the kernel"
+      exit 0
+    fi
+
+    # enable AppArmor enforcement
+    echo "Enable AppArmor enforcement of unprivileged user namespace restrictions"
+    sysctl -w kernel.apparmor_restrict_unprivileged_userns=1
+
+    # and remove the AppArmor permission from the snap's AppArmor profile
+    sed -i 's/userns,/##spread##userns,/' /var/lib/snapd/apparmor/profiles/snap.test-snapd-userns.with-userns-plug
+    tests.cleanup defer sed -i 's/^##spread##//' /var/lib/snapd/apparmor/profiles/snap.test-snapd-userns.with-userns-plug
+
+    # then reload the profile, ensuring the right apparmor_parser is used
+    APPARMOR_PARSER="apparmor_parser"
+    if snap debug sandbox-features --required apparmor:parser:snapd-internal; then
+      # TODO: don't use 3.0 abi as this has no knowledge of userns and so userns
+      # would still be implicitly allowed
+      APPARMOR_PARSER="/snap/snapd/current/usr/lib/snapd/apparmor_parser --config-file /snap/snapd/current/usr/lib/snapd/apparmor/parser.conf -b /snap/snapd/current/usr/lib/snapd/apparmor.d --policy-features /snap/snapd/current/usr/lib/snapd/apparmor.d/abi/3.0"
+    else
+      # ensure the system apparmor_parser doesn't use any pinned set of policy
+      # features
+      sed -i 's/^policy-features/##spread##policy-features/' /etc/apparmor/parser.conf
+      tests.cleanup defer sed -i 's/^##spread##//' /etc/apparmor/parser.conf
+    fi
+    $APPARMOR_PARSER -r /var/lib/snapd/apparmor/profiles/snap.test-snapd-userns.with-userns-plug
+
+    # running unshare -U as an unprivileged user to create a new user namespace
+    # should be denied
+    echo "When the userns AppArmor permission is removed creating a new user namespace as an unprivileged user is denied"
+    tests.session -u test exec sh -c "snap run test-snapd-userns.with-userns-plug -c '\$SNAP_COMMON/unshare -U id'" 2>&1 | MATCH 'unshare: unshare failed: Operation not permitted'
```

```suggestion
    tests.session -u test exec sh -c "snap run test-snapd-userns.with-userns-plug -c '\$SNAP_COMMON/unshare -U id'" 2>&1 | MATCH 'unshare: unshare: Permission denied'
```

from my tests I think the error message would be Permission denied. same for the test with root below

---

This is an ABI issue. snap does not add abi rules to the profiles it generates. So apparmor will either use the command line specified override, pinned abi set in ```/etc/apparmor/parser.conf``` or fallback to a default abi that corresponds roughly with Ubuntu from 2.13 era (its ~4.17 with net compat, and af_unix).

If snapd generates the a rule and the combined abi (kernel and policy) doesn't support it, the rule will get downgraded, dropped, or cause the compile to fail. In this case I think we are getting the ```info="same as current profile, skipping"``` message because the parser is dropping the rule because of the policy abi, so the compiled policy is the same (kernel does a bcmp for that message to be generated).

We are planning an additional parser mode for apparmor 4, which will allow the presence of a rule to override the policy abi (with a warning) as long as the rule is supported in the kernel. The kernel of course will still have to support the rule or it will be ignored.

---

I explicitly disable the use of the default pinned ABI though in parser.conf - https://github.com/snapcore/snapd/pull/12844/files#diff-305ae78661bbc1fbfb9069fcd04646de5be7720b148392d344089bccddfce8cfR102 - yet as @gegarcia notes, when the profile is then reloaded it does not appear to change - so for some reason apparmor_parser is still pinning the ABI even with `policy_features` being disabled.

---

yes. I think maybe a bit of miscommunication. I was merely enumerating how the parser should be determining the abi.

The important part is if the abi intersection (both policy and kernel) does not support user namespace, then the parser will drop the user namespace rule. And when the policy is loaded we can end up with the policy being the same.

Neither the default abi used for profiles that don't have abi rules nor the abi for 3.0 contain support for user namespace rules. So that portion of the profile when generated should be the same.

The other thing to understand about apparmor abi's is they are NOT a number they are a set of features. So they mix and match in ways that is different than having a number to compare against. So a profile built against two different abis could possibly actual end up being the same.

That is not to say there isn't a bug here. I need to look at the profile, and explicitly compile it with both the default and 3.0 abi. Something I haven't done yet.

---

I had mistakenly assumed that if the abi was not pinned, and if the profile did not specify the abi, then the kernel abi (set of features supported by the running kernel) would be used. Instead the default abi defined by the parser is used. If that's the expected behavior, then there is no bug.

@alexmurray for the test to work you will need to specify the abi in the profile, or while loading the profile like I did with `$APPARMOR_PARSER --policy-features=\<kernel\> -r ...`. Note that we don't currently provide an abi that has the userns feature defined in /etc/apparmor.d/abi/. So you can either use "kernel" or provide the abi file yourself. The tool `aa-features-abi` can help you.

---

In AppArmor 3+ the kernel ABI is never directly used unless directly specified. This is because under AppArmor 2.x the use of the kernel ABI by default when a pinned ABI was not present lead to kernel patches being reverted. To fix that we had to put the onus on policy devs to specify it directly. The compiled ABI is always an intersection of the kernel and policy specified ABIs. To get just the kernel ABI you specify kernel on the policy side

the policy ABI can specified by using any of
- policy rule ```abi=<kernel>```
- cmdline/config
  -  --policy-features=kernel
  -  --override-policy-feature=kernel    #required if you want to override an abi rule in the policy

alternately you can specify the "kernel" feature set to come from a file directly using
- cmdline/config
  - --features-file=<XXX>   # sets both policy features and kernel features
  - --kernel-features=<XXX> --policy-features=<XXX>

---

Thanks for the hints re `--policy-features=<kernel>` - I have used that for now in this test and it is working as expected \o/! So I think perhaps we should pursue this in general when using apparmor in snapd - ie. specify this along with `--Werror --Werror=rule-not-enforced` when probing for supported features in the parser so that we get an accurate reflection of what is supported by the parser *in combination with the kernel* - and perhaps we also need to store the kernel features in `system-key` so we can regenerate policy when the kernel gains new features. This won't work on 2.x apparmor (bionic + xenial) but in that case the behaviour would be the same as it is now which I think is still reasonable.